### DISCUSSION TITLE
What is webGPU support in Emscripten?

### DISCUSSION THREAD

**Discussion by yangfangfang1204** [emscripten-core/emscripten#15645] [Q&A] [answered]
In Emscripten release notes , v1.39.8: 02/14/2020
- WebGPU Compute fixes. Simple examples now work. See #10367.

so In Emscripten， What is webGPU support ？ support GPU compute ？ and how i use it to calculation of acceleration matrix ? if there is some demo ?

Another question : Whether Emscripten supports WebGPU rendering ? and how i use it to improved rendering performance ?

**Comment by yangfangfang1204** [comment]
Dawn is a C++ implementation of WebGPU used in Chromium. It can be used to target WebGPU in C and C++ applications that can then be ported to WebAssembly using Emscripten and automatically take advantage of WebGPU in the browser.

so the Dawn support webgpu renderPipeline ?

**Reply by kainino0x** [reply]
> so the Dawn support webgpu renderPipeline ?

Yes, when you use the C API for WebGPU render pipelines it will be translated to the browser's API.

**Comment by kainino0x** [comment, marked as answer]
WebGPU support in Emscripten is the implementation of `webgpu.h` (https://github.com/webgpu-native/webgpu-headers/blob/main/webgpu.h) on top of the browser-exposed WebGPU API (https://gpuweb.github.io/gpuweb/).

**WebGPU is still under development** and not yet exposed by default in browsers.
More info and sample links here: https://github.com/gpuweb/gpuweb/wiki/Implementation-Status#materials

JavaScript WebGPU samples: https://austin-eng.com/webgpu-samples/
Extremely minimal demo of an application using webgpu.h to target web through Emscripten: https://github.com/kainino0x/webgpu-cross-platform-demo

WebGPU support in Emscripten doesn't affect or enhance any applications that aren't using themselves using WebGPU directly. So you won't get improved rendering performance/features without adding engine support for a WebGPU rendering backend.

Hope this helps!

**Comment by FuXiii** [comment]
Where can find the document about `WebGPU` in `Emscripten` and some function like  `WGPUDevice emscripten_webgpu_get_device(void)` in `html5_webgpu.h` ?

(Ｔ▽Ｔ)

**Reply by kainino0x** [reply]
If you haven't seen it yet, this newly published article should be very helpful: https://developer.chrome.com/blog/webgpu-cross-platform/

`emscripten_webgpu_get_device` is deprecated. It's only for use with `preinitializedWebGPUDevice` which is going to be removed. It was needed in early prototypes because when we started we didn't have working implementations for requestAdapter/requestDevice in some cases, but it hasn't been a priority to clean up. (You can see what it does, it's quite simple:
https://github.com/search?q=repo%3Aemscripten-core%2Femscripten%20emscripten_webgpu_get_device&type=code)
